controllers/clustercache/cluster_accessor.go

```diff
@@ -495,3 +497,80 @@ func (ca *clusterAccessor) unlock(ctx context.Context) {
+// for the given Cluster.
+type WatchInput = TypedWatchInput[client.Object, ctrl.Request]
+
+// TypedWatchInput specifies the parameters used to establish a new watch for a workload cluster.
+// A source.TypedKind source (configured with Kind, TypedEventHandler and Predicates) will be added to the Watcher.
+// To watch for events, the source.TypedKind will create an informer on the Cache that we have created and cached
+// for the given Cluster.
+type TypedWatchInput[object client.Object, request comparable] struct {
+	// Name represents a unique Watch request for the specified Cluster.
+	// The name is used to track that a specific watch is only added once to a cache.
+	// After a connection (and thus also the cache) has been re-created, watches have to be added
+	// again by calling the Watch method again.
+	Name string
+
+	// Watcher is the watcher (controller) whose Reconcile() function will be called for events.
+	Watcher ScopedWatcher[request]
+
+	// Kind is the type of resource to watch.
+	Kind object
+
+	// EventHandler contains the event handlers to invoke for resource events.
+	EventHandler handler.TypedEventHandler[object, request]
+
+	// Predicates is used to filter resource events.
+	Predicates []predicate.TypedPredicate[object]
+}
+
+// NewWatcher creates a Watcher on the workload cluster.
+// A source.TypedKind source (configured with Kind, TypedEventHandler and Predicates) will be added to the ScopedWatcher.
+// To watch for events, the source.TypedKind will create an informer on the Cache that we have created and cached
+// for the given Cluster.
+func NewWatcher(input WatchInput) Watcher {
+	return &typedWatcher[client.Object, ctrl.Request]{
+		name:         input.Name,
+		kind:         input.Kind,
+		eventHandler: input.EventHandler,
+		predicates:   input.Predicates,
+		watcher:      input.Watcher,
+	}
+}
+
+// NewTypedWatcher creates a Watcher on the workload cluster.
+// A source.TypedKind source (configured with Kind, TypedEventHandler and Predicates) will be added to the ScopedWatcher.
+// To watch for events, the source.TypedKind will create an informer on the Cache that we have created and cached
+// for the given Cluster.
+func NewTypedWatcher[object client.Object, request comparable](input TypedWatchInput[object, request]) Watcher {
+	return &typedWatcher[object, request]{
+		name:         input.Name,
+		kind:         input.Kind,
+		eventHandler: input.EventHandler,
+		predicates:   input.Predicates,
+		watcher:      input.Watcher,
+	}
+}
+
+type typedWatcher[object client.Object, request comparable] struct {
+	name         string
+	kind         object
+	eventHandler handler.TypedEventHandler[object, request]
+	predicates   []predicate.TypedPredicate[object]
+	watcher      ScopedWatcher[request]
+}
+
+func (tw *typedWatcher[object, request]) Name() string     { return tw.name }
+func (tw *typedWatcher[object, request]) KindType() string { return fmt.Sprintf("%T", tw.kind) }
```

Comment:
Let's call this `Object() client.Object` and return tw.kind directly

In the first few cases we can then use %T in the errors and in l.436++ we can do this to get the Kind cleanly
```go
apiutil.GVKForObject(ca.lockedState.connection.cachedClient.Scheme(), watcher.Object())
```

Reply:
I like the `Object()` approach. With that can't we just do `%T` everywhere? Why would we need `apiutil.GVKForObject()`? Is it to avoid reflection?

Reply:
I think %T prints *Node instead of Node and it assumes that the Go type is identical to the kind (which is usually the case but it doesn't have to be)

Reply:
We can keep it as is. All good as long as we don't pretend it is the kind :)